mm/map/mm_map.c

```diff
@@ -0,0 +1,337 @@
+ *
+ ****************************************************************************/
+
+void mm_map_destroy(FAR struct mm_map_s *mm)
+{
+  FAR struct mm_map_entry_s *map;
+
+  while ((map = (FAR struct mm_map_entry_s *)sq_remfirst(&mm->mm_map_sq)))
+    {
+      /* Pass null as group argument to indicate that actual MMU mappings
+       * must not be touched. The process is being deleted and we don't
+       * know in which context we are. Only kernel memory allocations
+       * need to be freed by drivers
+       */
+
+      if (map->munmap)
+        {
+          /* Unmap the whole region */
+
+          if (map->munmap(NULL, map, (void *)map->vaddr, map->length) < 0)
+            {
+              /* This is a bug. A driver provides munmap, but can't munmap
+               * the region. Just delete the map entry.
+               */
+
+              kmm_free(map);
+            }
+        }
+      else
+        {
+          /* Mappping has been added, but there is no unmap. Just delete the
+           * map entry.
+           */
+
+          kmm_free(map);
+        }
+    }
+
+  nxmutex_destroy(&mm->mm_map_mutex);
+}
+
+/****************************************************************************
+ * Name: mm_map_add
+ *
+ * Description:
+ *   Add a mapping to task group's mm_map list
+ *
+ ****************************************************************************/
+
+int mm_map_add(FAR struct mm_map_entry_s *entry)
+{
+  FAR struct tcb_s *tcb = nxsched_self();
+  FAR struct task_group_s *group = tcb->group;
+  FAR struct mm_map_s *mm = &group->tg_mm_map;
+  FAR struct mm_map_entry_s *map = kmm_zalloc(sizeof(struct mm_map_entry_s));
+  int ret;
+
+  if (!entry)
+    {
+      return -EINVAL;
+    }
+
+  if (!map)
+    {
+      return -ENOMEM;
+    }
+
+  /* Copy the provided mapping and add to the list */
+
+  *map = *entry;
+
+  ret = nxmutex_lock(&mm->mm_map_mutex);
+  if (ret < 0)
+    {
+      kmm_free(map);
+      return ret;
+    }
+
+  sq_addfirst((sq_entry_t *)map, &mm->mm_map_sq);
+
+  nxmutex_unlock(&mm->mm_map_mutex);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: mm_map_next
+ *
+ * Description:
+ *   Returns the next mapping in the list.
+ *
+ ****************************************************************************/
+
+FAR struct mm_map_entry_s *mm_map_next(
+                           FAR const struct mm_map_entry_s *entry)
+{
+  FAR struct tcb_s *tcb = nxsched_self();
+  FAR struct task_group_s *group = tcb->group;
+  FAR struct mm_map_s *mm = &group->tg_mm_map;
+  FAR struct mm_map_entry_s *map = NULL;
+
+  if (nxmutex_lock(&mm->mm_map_mutex) == OK)
+    {
+      if (entry == NULL)
+        {
+          map = (struct mm_map_entry_s *)sq_peek(&mm->mm_map_sq);
+        }
+      else
+        {
+          map = (struct mm_map_entry_s *)sq_next(((sq_entry_t *)entry));
+        }
+
+      nxmutex_unlock(&mm->mm_map_mutex);
+    }
+
+  return map;
```

Comment:
the lock doesn't fix any race condition, it's always unsafe to hold an entry without lock

Reply:
the lock protects th sq modifications. In a multi-threaded application separate threads can do mmap/munmap simultanously. Other parts of the single map_entry don't need to be protected. The queue modifications are only done in these functions, and that's why the mutex is there.

Reply:
> the lock tries to protect the sq modifications. In a multi-threaded application separate threads can do mmap/munmap simultanously.
> 

Yes, but it's unsafe to return the entry after locking. The caller of mm_map_find_contains which touch the returned value will suffer the race condition.

> But it is right that since mm_map_remove takes a direct reference to map, it needs to be locked also above in some cases to protect the queue.
> 
> I'll fix the locking

Reply:
yes, fixed in mm/shm driver and in fs_munmap. Changed the mutex to rmutex, keep it locked always when managing the list. The concept is very similar to inode handling.